vgrabbj.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH VGRABBJ 1 "February  24, 2002"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
vgrabbj \- grab images from a v4l-compatible webcam
.SH SYNOPSIS
.B vgrabbj
.RI [ options ]
.br
.SH DESCRIPTION
This manual page documents briefly the
.B vgrabbj
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBvgrabbj\fP is a program that will grab images from any v4l-capable 
device which supports one of the rgb24, rgb32, yuv420, yuv420p, yuyv, 
uyvy, or yuv422 palettes and saves the image as a .jpg, .png, or .pnm file. 
.PP
Optional, it can timestamp the resulting image, and/or upload it to 
a ftp-server. Additionally, a daemon-mode is available to do the 
above regularly.
.PP
On startup, the default configuration file \fI/etc/vgrabbj.conf\fP is
read and evaluated. If a configuration file is specified via the
\fB-c\fP parameter (see below), that file is read and parsed. Values
which are present in the default config file keep their values.
Finally, the commandline is evaluated. If an option is presented 
more than once, the last in the commandline is taken.
If no options given, vgrabbj tries to open the default v4l-device
\fI/dev/video0\fP, grab an image and return it on \fI/dev/stdout\fP.
So be aware.
.SH OPTIONS
All options are either numeric values, strings, or booleans. If it
is a boolean switch (e.g. no value/string shown after the switch),
it plainly switches the current setting of that particular option. 
Remember, vgrabbj.conf is read first, so any option set to on or 
off there will be toggled by the commandline switch.
.TP
\fB\-h\fP
Show summary of options and version number.
\fB\-c\fP <\fIfilename\fP>
Specifies \fIfilename\fP as alternative configuration file.
.TP
\fB\-l\fP <\fIseconds\fP> 
Daemonize & loop with \fIseconds\fP intervals (min. 1).
.TP
\fB\-L\fP <\fImicroseconds\fP> 
Daemonize & loop \fImicroseconds\fP intervals (min. 1). Same as \fB\-l\fP
except you can control it more precisely.
.TP
\fB\-a\fP 
Enable vgrabbj's brightness adjustment (default: disabled). To use \fBvgrabbj\fP's
brightness adjustment, RGB24 support of the device is necessary. You may
have to provide it to your device via the \fB-F\fP parameter (see below).
If your device does not support RGB24, this setting is silently ignored.
.TP
\fB\-q\fP <\fIquality\fP> 
Sets the quality of a jpeg output to \fIquality\fP (1-100, default: 75).
.TP
\fB\-i\fP <\fIimagesize\fP>
Sets the imagesize of input device, where \fIimagesize\fP is one of: 
.IP
.nf
.ta
 sqcif= 128x96,     qsif = 160x120, 
 qcif = 176x144,    sif  = 320x240, 
 cif  = 352x288,    vga  = 640x480, 
 svga = 800x600,    xga  = 1024x768, 
 sxga = 1280x1024,  uxga = 1600x1200, 
.fi
.IP
(default: cif)
.TP
\fB\-w\fP <\fIwidth\fP>
Sets the width of the image to \fIwidth\fP. Needs \fB\-H\fP.
The v4l-device has to support the image width you set with
this option.
.TP
\fB\-H\fP <\fIheight\fP>
Sets the height of the image to \fIheight\fP. Needs \fB\-w\fP.
The v4l-device has to support the image height you set with
this option.
.TP 
\fB\-o\fP <\fIformat\fP> 
Sets the output to \fIformat\fP (either \fBjpeg\fP, \fBpng\fP, or 
\fBpnm\fP, default: jpeg)
.TP
\fB\-f\fP <\fIfilename\fP>
Output is written to \fIfilename\fP (default: \fI/dev/stdout\fP)
.TP
\fB\-A\fP <\fIpath+filename\fP>
Additional to the output of \fB-f\fP another file is written to the
named \fIpath+filename\fP, whereas the whole may be formatted
according to \fBstrftime\fP(2). This enables the creation of an archive
of images. Currently, the output is only supported on the same
filesystem as the original output.
.TP
\fB\-E\fP <\fIvalue\fP>
Defines how many images shall be read until a copy for the archive
is made. E.g. 1 means each image taken is also put in the archive.
5 would mean every fifth image taken is put in the archive.
.TP
\fB\-M\fP <\fIvalue\fP>
Sets the maximum number of images kept in the archive. If you defined
5 seconds for daemon mode, archive every image, you collect 720 images
per hour. If you set this switch to 720 you would keep an hour of images
in the archive.
.TP
\fB\-d\fP <\fIdevice\fP> 
Sets the input device to \fIdevice\fP (default: \fI/dev/video0\fP)
.TP
\fB\-s\fP <\fIdevice\fP>
This flag causes vgrabbj to print the v4l capabilities of <\fIdevice\fP>
and exit. 
.TP
\fB\-C\fP 
Causes vgrabbj to open the device only while image data is grabbed. 
This is helpful if you need access to the device besides vgrabbj. It's
use is discouraged due to possible problems with memory fragmentation.
Default is to keep the v4l-device open while \fBvgrabbj\fP is running.
.TP
\fB\-g\fP
Disables setting of the image size (see \fB\-i\fP, \fB\-H\fP, and 
\fB\-w\fP. Necessary for certain v4l-devices which do not support to be
set to a particular image size (e.g. IBM-usb-cam, QuickCam). For these
devices you have to specify the size at module load time.
.TP
\fB\-S\fP
Switches colormap from BGR to RGB (or vice versa). Use if you have odd
colors.
.TP
\fB\-R\fP
Swaps left and right of the image like a mirror.
.TP
\fB\-U\fP
Swaps top and bottom of the image like a mirror (e.g. for upside-down
montage of the cam).
.TP
\fB\-G\fP
Turns off use of mmap'ed memory. see \fIvgrabbj.conf\fP(5) for details.
.TP
\fB\-D\fP <\fIvalue\fP>
Sets the loglevel of \fBvgrabbj\fP. Valid for \fIvalue\fP are 0, 2, 3, 4, 6, 7
(default: 4). 0 means no messages at all, 7 is debug. Don't use 7 if
you are running \fBvgrabbj\fP in daemon mode, it'll surely fill your
syslog. Be aware, that loglevel 7 will also write your ftp-password
(if defined in \fBvgrabbj.conf\fP(5)) in the logfile!
.TP
\fB\-n\fP
Disable the use of the temporary outputfile and write directly to the
output-file specified with \fB-f\fP. Output to \fI/dev/stdout\fP is always
written directly without the use of a temporary file.
.sp
If a \fIfilename\fP different from \fI/dev/stdout\fP was given with the \fB-f\fP
parameter, \fBvgrabbj\fP defaults to use a temporary file called \fIfilename\fP.tmp
to write the image. This file is then renamed to \fIfilename\fP to shorten the
time period corrupt data may be read from \fIfilename\fP.
.TP
\fB\-F\fP <\fIpalette\fP>
Force the usage of the specified \fIpalette\fP. If \fIpalette\fP is not
supported by either \fBvgrabbj\fP or the v4l-device, fallback to a palette supported
by both is provided. If no common denominator is available, \fBvgrabbj\fP exits. See
\fBDESCRIPTION\fP for details on palettes supported by \fBvgrabbj\fP. Possible values
are (copy from videodev.h):
.IP
.nf
.ta
 1  VIDEO_PALETTE_GREY      Linear greyscale
 2  VIDEO_PALETTE_HI240     High 240 cube (BT848)
 3  VIDEO_PALETTE_RGB565    565 16 bit RGB
 4  VIDEO_PALETTE_RGB24     24bit RGB
 5  VIDEO_PALETTE_RGB32     32bit RGB
 6  VIDEO_PALETTE_RGB555    555 15bit RGB
 7  VIDEO_PALETTE_YUV422    YUV422 capture
 8  VIDEO_PALETTE_YUYV
 9  VIDEO_PALETTE_UYVY
 10 VIDEO_PALETTE_YUV420
 10 VIDEO_PALETTE_YUV411    YUV411 capture
 11 VIDEO_PALETTE_RAW       RAW capture (BT848)
 12 VIDEO_PALETTE_YUV422P   YUV 4:2:2 Planar
 13 VIDEO_PALETTE_YUV411P   YUV 4:1:1 Planar
 14 VIDEO_PALETTE_YUV420P   YUV 4:2:0 Planar
 15 VIDEO_PALETTE_YUV410P   YUV 4:1:0 Planar
 16 VIDEO_PALETTE_PLANAR
 17 VIDEO_PALETTE_COMPONENT
.fi
.IP
(no default, depends on v4l-device)
.TP
\fB\-z\fP <\fIvalue\fP>
\fIvalue\fP images are grabbed and immediately discarded. Only the
last image is written to the output. This might be helpful if your v4l-device
needs to take a couple of images to auto-adjust to the current light.
.TP
\fB\-X\fP
Does not fork daemon process and keeps vgrabbj in daemon mode in foreground,
attached to the terminal. Useful for debugging purpose. Not available in
config file.
.SH HARDWARE CONTROL OPTIONS
.sp
With the following options you can control the hardware settings of your
cam - if it is supported by your cam. Be aware that these options might
break your hardware, as there is no check on what values are available,
allowed, and supported by your hardware.
.TP
\fB\-W\fP <\fIvalue\fP>
Whiteness definition.
.TP
\fB\-r\fP <\fIvalue\fP>
Color definition.
.TP
\fB\-x\fP <\fIvalue\fP>
Contrast definition.
.TP
\fB\-b\fP <\fIvalue\fP>
Brightness definition.
.TP
\fB\-u\fP <\fIvalue\fP>
Hue definition.
.SH TIMESTAMP OPTIONS
.sp
The following options are available only if \fBvgrabbj\fP was compiled with
the freetype library available. If you set any one of these options on the
commandline timestamp will be enabled with default values for all other
options.
.TP
\fB\-e\fP
Switches timestamp to be used or not. Default is off, but may be overridden
by config-file, so just try to enable this switch if there is no timestamp
in the resulting image.
.TP
\fB\-t\fP <\fIfilename\fP> 
The TrueType font to be used for the timestamp
(default \fI/usr/X11R6/lib/X11/fonts/TrueType/Arialn.ttf\fP)
.TP
\fB\-T\fP <\fIsize\fP>
Set the fontsize to \fIsize\fP for timestamp (3-100, default: 12).
.TP
\fB\-p\fP "<\fIformat-str\fP>"
Defines the timestamp. \fIformat-str\fP can be plain text or any
strftime format and has to be enclosed with '""' (default: "%a, %e.
%B %Y \- %T" \- see \fBstrftime\fP(3) for details)
.TP
\fB\-P\fP <\fIvalue\fP>
Alignment of the timestamp in the image. Possible 
\fIvalue\fP:
.IP
.nf
.ta
0 = upper left,    1 = upper right,
2 = lower left,    3 = lower right,
4 = upper center,  5 = lower center
.fi
.IP
(default: 1).
.TP
\fB\-m\fP <\fIvalue\fP>
Defines the blend between font background and image (1-100, default: 60).
.TP
\fB\-B\fP <\fIvalue\fP>
\fIvalue\fP pixels will be used as border around the timestamp string (1-255,
default: 2).
.SH FTP OPTIONS
.P
There are no options for ftp-upload on the command line. This is due to the need
to provide a password which would be visible via the \fBps\fP(1) command. See
\fBvgrabbj.conf\fP for details on how to set-up the ftp-upload option.
Additionally, vgrabbj has to be compiled with libftp support.
.SH BUGS
.P
%F in the archive time-string (year with century) causes memory corruption at a
pentium-machine running vgrabbj (not validated). No further bugs are known at this time.
.SH SEE ALSO
\fBvgrabbj.conf\fP(4), \fBstrftime\fP(3), \fBps\fP(1)
.SH AUTHOR
This manual page was originally written by Michael Janssen 
<janssen@cns.uni.edu>, for the Debian GNU/Linux system and enhanced by 
Jens Gecius <devel@gecius.de>.


vgrabbj/ChangeLog
2013-09-04  Ludovic Rousseau <ludovic.rousseau@free.fr>

	* release 0.9.7
	* integrate the Debian changes upstream
	* fix compiler warnings

2004-01-03  Jens Gecius  <jens@gecius.de>

	* vgrabbj.conf.5/vgrabbj.1: 
			Added all missing options to the
			manuals, corrected misspelled
			options. Certain clarifications.
	* vgrabbj.c:	Added option -X to let daemon
			mode run in foreground (e.g. not
			detaching from current terminal)

2003-12-31  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Changed pointer-update to be
			correct in any case. Fixed error
			in archiving functionality.

2003-10-06  Robert van der Meulen <rvdm@wiretrip.org>
	* General:	Added support for passive ftp.
 
2003-09-10  Jens Gecius  <jens@gecius.de>
	* various	Included a patch from Koos van den
			Hout and Arthur van Leeuwen to
			enable vgrabbj to flip top to
			bottom. This enables upside-down
			mounting, e.g. on a ceiling.
	* various	Included a patch from David Austin
			to enable vgrabbj to take detailed
			control of the cams hardware 
			settings (e.g. brightness-adjustment,
			hue, etc.).
			This one needs further testing and
			adjustment to the options setup of
			previous versions. As of now this
			version might break your setup.

2003-07-28  Jens Gecius  <jens@gecius.de>
	* v_conf.c:	Fixed typo in help screen
			(thanks to Tobias Wolter who filed
			this bug to DBTS)
			Fixed bug causing segfault if not
			all optional libraries are present
			(thanks to Pascal Brisset for the
			patch! Sorry it took so long to
			apply it)

2002-10-15  Jens Gecius  <jens@gecius.de>

	* vgrabbj.h: 	Added option nousemmap needed for
			certain cams not switchable to
			palette RGB24 but unable to use
			mmap'ed memory.

	* v_options.h:	Fixed a small bug for ForcePalette.
			Improved img_size calculation.

2002-10-14  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Fixed nasty bug causing boolean
			commandline options not being parsed
			correctly.
			Added switch SwapRL (-R) to mirror
			the resulting image left-right.

2002-04-12  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Fixed bug for commandline switches
			on boolean values (segfault)
			Fixed possible wrong treatment of
			empty lines in conf

2002-04-02  Jens Gecius  <jens@gecius.de>

	* 		Final touch up, release 0.9.0
			with archive functionallity.

2002-03-29  Jens Gecius  <jens@gecius.de>

	* v_config.c:	rewrote/enhanced archive handling
			archivemax/archiveeach to manage
			archived pictures added.

2002-03-27  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	(after release) Fixed bug in ttf
			init. Changed internals slightly.

	* General: 	Update manuals, uploaded man's to
			cvs, increased version no. to 0.8.9,
			release.

2002-03-26  Jens Gecius  <jens@gecius.de>

	* vgrabbj.c:	Added Archive option to create an
			archive of images.
			Warning: no upper limit for archive
			implemented, yet. This option might
			fill your harddrive!!
			Currently works only if archive is
			on the same filesystem as the temp-
			output (same restriction for output
			and tmpout - they have to be on the
			same filesystem)
	
	* v_config.c: 	Fixed severe bugs in ftp-code:
			1. remoteDir was not read from conf
			2. if remoteDir not set screwy things
			might have happened...

2002-02-24  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Fixed missing explanation of parameter
			-c on the commandline for the -h
			commanline parameter.

			Enhanced manual.

2002-02-18  Jens Gecius  <jens@gecius.de>

	* vgrabbj.c: 	Finished rewrite of conf-parser. 
			(Almost) identical routines are now 
			used for command-line and conf-file.
			Further information is given in case
			of wrong options.
			Cleanup of code (partially), blow-up
			on some other areas (especially the
			conf-var has to be cleaned up now...)
			Well, ongoing progress, short of
			release 0.8.8 (testing).
			OpenOnce defaults ot On now (to avoid
			memory fragmentation and usb-messages
			in syslog). Use Off only if you need
			to access the device inbetween vgrabbj
			taking frames.

2002-02-16  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Further rewrite of conf-parser (not
			finished - sigh)
			Definitions of options and certain
			detailed definitones moved to structs
			(former: in code)
			Timestamp string does not need to be in
			paranthesis (conf-file) any more (but
			can - it will be filtered!)

2002-02-12  Jens Gecius  <jens@gecius.de>

	* vgrabbj.c: 	Changed handling of open-device in case
			of openonce (it is really opened only once
			now), avoiding lot's of kernel-uhci-msgs
			on my box. Hope you guys benefit, too.
			Again, I strongly recommend to use OpenOnce

2002-02-11  Jens Gecius  <jens@gecius.de>

	* vgrabbj.c: 	More split-up of source to make it more
			readable. Re-organization of mmap (now
			close to device open to prevent memory
			fragmentation. If the device is not
			otherwise needed, I strongly advise to 
			use OpenOnce yes (now default).
			Changed handling of Timestamp, Font-eninge
			is loaded/initialized once only at startup
			and if config is re-read.

2002-02-10  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Added DaemonSeconds as alternative to Daemon
			in conf - value given is taken as seconds
			instead of milliseconds (see conf).

	* vgrabbj.c: 	Added some exit routine to mmap'ed capture
			to avoid endless loops (2).

2002-02-09  Jens Gecius  <jens@gecius.de>

	* v_config.c:	Added a parameter to discard frames before
			the actual picture is taken. Might be
			necessary, if your cam needs a little time
			for adjustment of the current brightness.

2002-02-08  Jens Gecius  <jens@gecius.de>

	* v_config.c:	Fixed a #define bug in checkfiles (if you
			do not have TTF on your system it's
			relevant, otherwise not)
			
	* configure.in:	Certain systems seem to have the freetype
			header (v1) in a different directory than
			debian does. Added appropriate check.

2002-02-06  Jens Gecius  <jens@gecius.de>

	* v_config.c: 	Closed a memory hole generated by the

vgrabbj/vgrabbj.c
/* Simple Video4Linux image grabber. Made for my Philips Vesta Pro
 *
 * Copyright (C) 2000, 2001, 2002 Jens Gecius, Hannover, Germany
 * eMail: devel@gecius.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at you option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston MA 02111-1307,
 * USA
 */

#include <vgrabbj.h>
#include <v_plist.h>

int signal_terminate=0;

void cleanup(struct vconfig *vconf) {
  int tmp=vconf->loop;

  v_update_ptr(vconf);
  if (vconf->openonce) {
    free_mmap(vconf);
    close_device(vconf);
    }
#ifdef LIBTTF
  if ( vconf->use_ts && vconf->ttinit ) {
    Face_Done(vconf->ttinit->instance, vconf->ttinit->face);
    TT_Done_FreeType(vconf->ttinit->engine);
  }
  if ( vconf->ttinit ) {
    vconf->ttinit->properties=free_ptr(vconf->ttinit->properties);
    vconf->ttinit=free_ptr(vconf->ttinit);
  }
  vconf->timestamp=free_ptr(vconf->timestamp);
  vconf->font=free_ptr(vconf->font);
  v_error(vconf, LOG_DEBUG, "TrueType engine ended, vars freed");
#endif
  vconf->buffer=free_ptr(vconf->buffer);
  vconf->o_buffer=free_ptr(vconf->o_buffer);
  v_error(vconf, LOG_DEBUG, "Buffers freed");
  vconf->tmpout=free_ptr(vconf->tmpout);
  vconf->in=free_ptr(vconf->in);
  vconf->out=free_ptr(vconf->out);
  vconf->conf_file=free_ptr(vconf->conf_file);
  v_error(vconf, LOG_DEBUG, "vars freed");

  vconf=free_ptr(vconf);

  if ( tmp ) {
     closelog();
    _exit(1);
  }
}

/* Event handler for SIGHUP */
/* to re-read the configuration file */



void v_error(struct vconfig *vconf, int msg, const char *fmt, ...)
{
  va_list arg_ptr;
  static char buf[MAX_ERRORMSG_LENGTH];

  /* Loglevels:

     0     silent (NONE)
     2     severe errors which cause vgrabbj to exit in daemon mode (LOG_CRIT)
     3     critical errors which cause vgrabbj to exit in non-daemon mode
           and retry forever in daemon-mode (with a sleep(1) inbetween) (LOG_ERR)
     4     warning conditions (LOG_WARNING)
     6     normal information (LOG_INFO)
     7     debug information (LOG_DEBUG)

     Every log-message with a loglevel equal-smaller than LOGLEVEL-Level (supplied
     via command line, otherwise default=4) will be
         a) displayed on screen (stderr) if non-daemon
         b) written to syslog if daemon

     One exception:

     In daemon mode, every message before successful fork into background
     will be written to stderr (if any).

  */

  if ( msg <= vconf->debug ) {

    va_start(arg_ptr, fmt);
    vsprintf(buf, fmt, arg_ptr);
    strcat(buf, "\n");
    va_end(arg_ptr);
    if (vconf->loop && vconf->init_done) {
      syslog(msg, "%s", buf);
      if (msg == 3) {
	sleep(1);
      }
    } else {
      fprintf(stderr, "%s", buf);
      fflush(stderr);
    }

  }

  /* Decision about exiting is independent of log or not to log */

  if ( (vconf->loop && msg < 3) || (msg==3 && vconf->err_count++ > 3600) ) {
    syslog(LOG_ERR, "Fatal Error (daemon), exiting...\n"); // exit
    cleanup(vconf);
    closelog();
    _exit(1);
  } else if ( !vconf->loop && msg < 4 ) {
    fprintf(stderr, "Fatal Error (non-daemon), exiting...\n"); // exit
    cleanup(vconf);
    exit(1);
  }
}

/* Get information from v4l device and show them  */

void show_capabilities(char *in, char *pname)
{
  struct video_capability cap;
  struct video_window win;
  struct video_picture pic;
  int dev;

  if ( (dev = v4l1_open(in, O_RDONLY)) < 0 ) {
    fprintf(stderr, "Can't open device %s\n", in);
    exit(1);
  }
  if (v4l1_ioctl(dev, VIDIOCGCAP, &cap) < 0) {
    fprintf(stderr, "Can't get capabilities of device %s\n", in);
    exit(1);
  }
  if (v4l1_ioctl(dev, VIDIOCGPICT, &pic) < 0) {
    fprintf(stderr, "Can't get picture properties of device %s\n", in);
    exit(1);
  }
  if (v4l1_ioctl(dev, VIDIOCGWIN, &win) < 0) {
    fprintf(stderr, "Can't get overlay values of device %s\n", in);
    exit(1);
  }

  fprintf(stderr,"%s, Version %s\n"
	  "Videodevice name: %s (%s)\n"
	  "Capabilities\n"
	  "Type     : %d\tValues can be looked up at videodev.h\n"
	  "Channels : %d\n"
	  "Audio    : %d\n"
	  "MaxWidth : %d\n"
	  "MaxHeight: %d\n"
	  "MinWidth : %d\n"
	  "MinHeigth: %d\n"
	  "\nCurrent Settings:\n"
	  "Brightness: %d\n"
	  "Hue       : %d\n"
	  "Color     : %d\n"
	  "Contrast  : %d\n"
	  "Whiteness : %d\n"
	  "Depth     : %d\n"
	  "Palette   : %s (%d)\n"
	  "Width     : %d\n"
	  "Height    : %d\n"
	  "Chromakey : %d\n",
	  basename(pname), VERSION, in, cap.name, cap.type, cap.channels, cap.audios,
	  cap.maxwidth, cap.maxheight, cap.minwidth, cap.minheight,
	  pic.brightness, pic.hue, pic.colour, pic.contrast,
	  pic.whiteness, pic.depth, plist[pic.palette].name, pic.palette,
	  win.width, win.height, win.chromakey);
  dev=v4l1_close(dev);
  if (dev)
    fprintf(stderr, "Error occured while closing %s\n", in);
  exit(0);
}


/* Read the image from device, adjust brightness, if wanted, for RGB24 */

static char *read_image(struct vconfig *vconf, int size) {
  int f, newbright;
  int err_count;
  int discard=vconf->discard;

  v_error(vconf, LOG_DEBUG, "Palette to be used: %s (%d), size: %d",
	  plist[vconf->vpic.palette].name, vconf->vpic.palette, img_size(vconf, vconf->vpic.palette));

  /* Opening input device */
  if ( !vconf->openonce ) {
    open_device(vconf);

    /* and Re-initialize the palette, in case someone changed it meanwhile */

    while (v4l1_ioctl(vconf->dev, VIDIOCSPICT, &vconf->vpic) < 0 )
      v_error(vconf, LOG_ERR, "Device %s couldn't be reset to known palette %s",
	      vconf->in, vconf->vpic.palette);
    if (vconf->windowsize)
      while (v4l1_ioctl(vconf->dev, VIDIOCSWIN, &vconf->win) )
	v_error(vconf, LOG_ERR, "Problem setting window size"); // exit

    set_picture_parms(vconf);
  }


  /* Read image via read() */

  if (!vconf->usemmap) {
    if (vconf->autobrightness)
      v_error(vconf, LOG_INFO, "Forced to use brightness adj. - using read()");
    else
      v_error(vconf, LOG_INFO, "Could not get mmap-buffer - Falling back to read()");
    do {
      err_count=0;
      if (vconf->autobrightness && vconf->vpic.palette==VIDEO_PALETTE_RGB24) {
	v_error(vconf, LOG_INFO, "Doing brightness adjustment");
	do {
	  while (v4l1_read(vconf->dev, vconf->buffer, size) < size)
	    v_error(vconf, LOG_ERR, "Error reading from %s", vconf->in);
	  f = brightness_adj(vconf, &newbright);
	  if (f) {
	    vconf->vpic.brightness += (newbright << 8);
	    if (v4l1_ioctl(vconf->dev, VIDIOCSPICT, &vconf->vpic)==-1)
	      v_error(vconf, LOG_WARNING, "Problem setting brightness");
	    err_count++;

	    if (err_count>100) {
	      v_error(vconf, LOG_WARNING, "Brightness not optimal");
	      break;
	    }
	  }
	} while (f);
	v_error(vconf, LOG_INFO, "Brightness adjusted");
      } else {
	v_error(vconf, LOG_DEBUG, "Using normal read for image grabbing");
	v4l1_read(vconf->dev, vconf->buffer, size);
      }
    } while (discard--);

    /* We're reading the image via a mmap'd area of the driver */

  } else {
    v_error(vconf, LOG_DEBUG, "Using mmap for image grabbing");
    if (!vconf->openonce)
      init_mmap(vconf);
    do {
      err_count=0;
      do {
	if  (err_count++>100) {
	  v_error(vconf, LOG_ERR, "Could not grab frame (100 tries)");
	  break;
	}
      } while (v4l1_ioctl(vconf->dev, VIDIOCMCAPTURE, &vconf->vmap) < 0);

      err_count=0;
      do {
	if (err_count++>100) {
	  v_error(vconf, LOG_ERR, "Could not sync with frame (100 tries)");
	  break;
	}
      } while (v4l1_ioctl(vconf->dev, VIDIOCSYNC, &vconf->vmap.frame) < 0);

      vconf->buffer=memcpy(vconf->buffer, vconf->map+vconf->vbuf.offsets[vconf->vmap.frame], size);

      if (discard)
	v_error(vconf, LOG_DEBUG, "%d frames to discard", discard);

    } while (discard--);
    if (!vconf->openonce)
      free_mmap(vconf);
  }

  v_error(vconf, LOG_DEBUG, "Image successfully read");

  if (!vconf->openonce)
    close_device(vconf);

  return vconf->buffer;
}


static char *conv_image(struct vconfig *vconf) {

  unsigned long i,j;

  switch (vconf->vpic.palette) {
  case VIDEO_PALETTE_RGB24:
    vconf->o_buffer=memcpy(vconf->o_buffer, vconf->buffer,
		    vconf->win.width * vconf->win.height * 3);
    v_error(vconf, LOG_DEBUG, "No conversion, we have RGB24");
    break;
  case VIDEO_PALETTE_GREY:
    for(i=0 ;i<(vconf->win.width * vconf->win.height * 1) ;i++)
    {
	for(j=0 ;j<3 ;j++)
	{
	    vconf->o_buffer[(3*i)+j] = vconf->buffer[i];
	}
    }
    v_error(vconf, LOG_DEBUG, "Got GREY, converting...");
    break;
  case VIDEO_PALETTE_RGB32:
    v_error(vconf, LOG_INFO, "Got RGB32, converting...");

    vconf->o_buffer=conv_rgb32_rgb24(vconf);
    break;
  case VIDEO_PALETTE_YUV420P:
    v_error(vconf, LOG_INFO, "Got YUV420p, converting...");

    ccvt_420p_bgr24(vconf->win.width, vconf->win.height, vconf->buffer,
		    vconf->buffer + (vconf->win.width * vconf->win.height),
		    vconf->buffer + (vconf->win.width * vconf->win.height)+
		    (vconf->win.width * vconf->win.height / 4),
		    vconf->o_buffer);
    break;
  case VIDEO_PALETTE_YUV420:
    v_error(vconf, LOG_INFO, "Got YUV420, converting...");

    ccvt_420i_rgb24(vconf->win.width, vconf->win.height, vconf->buffer, vconf->o_buffer);
    break;
  case VIDEO_PALETTE_YUYV:
  case VIDEO_PALETTE_YUV422:
    v_error(vconf, LOG_INFO, "Got YUYV, converting...");

    ccvt_yuyv_bgr24(vconf->win.width, vconf->win.height, vconf->buffer, vconf->o_buffer);
    break;
  case VIDEO_PALETTE_UYVY:
    v_error(vconf, LOG_INFO, "Got UYVY, converting...");

    ccvt_uyvy_bgr24(vconf->win.width, vconf->win.height, vconf->buffer, vconf->o_buffer);
    break;
  default:
    v_error(vconf, LOG_CRIT, "Should not happen - Unknown input image format");
    break;
  }

  if (vconf->vpic.palette!=VIDEO_PALETTE_RGB24)
    v_error(vconf, LOG_DEBUG, "converted to RGB24");

  if (vconf->switch_bgr) {
    vconf->o_buffer=switch_color(vconf);
    v_error(vconf, LOG_DEBUG, "Switched from BGR to RGB - or vice versa");
  }
  return vconf->o_buffer;
}


/* Main loop  */

int main(int argc, char *argv[])
{
  struct vconfig *vconf=NULL;

#ifdef DEBUGGING
  mtrace();
#endif

  vconf=v_init(vconf, argc, argv);

  if (vconf->loop && (!vconf->nofork) )
    daemonize(vconf, basename(argv[0]));
  else
    v_error(vconf, LOG_WARNING, "Reading image from %s", vconf->in);

  // now start the loop (if daemon), read image and convert it, if necessary
  do {
    vconf->buffer=read_image(vconf, img_size(vconf, vconf->vpic.palette));
    vconf->o_buffer = conv_image(vconf);

    if (vconf->swaprl)
      vconf->o_buffer=swap_left_right(vconf->o_buffer, vconf->win.width, vconf->win.height);

    if (vconf->swaptb)
      vconf->o_buffer=swap_top_bottom(vconf->o_buffer, vconf->win.width, vconf->win.height);

#ifdef LIBTTF
    if (vconf->use_ts)
      vconf->o_buffer=inserttext(vconf->ttinit, vconf->o_buffer, vconf);
#endif

    write_image(vconf);

    vconf->err_count=0;
    usleep(vconf->loop);

    // We got a signal, we do have to...
    switch (signal_terminate) {
    case 0:
      // No signal received
      break;
    case SIGTERM:
      // We got sigkill, cleanup (for daemon mode) and exit
      signal_terminate=0;
      cleanup(vconf);
      break;
    case SIGHUP:
      // We got sighup, re-read the config-file (do NOT parse commandline)
      signal_terminate=0;
      vconf=v_reinit(vconf);
      break;
    default:
      v_error(vconf, LOG_DEBUG, "Unknown signal received, ignoring.");
      break;
    }
  } while (vconf->loop);

  cleanup(vconf);
  exit(0);
}






